cpu/sam0_common/periph/i2c.c

```diff
@@ -329,48 +422,105 @@ void _i2c_poweroff(i2c_t dev)
 static int _i2c_start(SercomI2cm *dev, uint16_t addr)
 {
     /* Wait for hardware module to sync */
-    DEBUG("Wait for device to be ready\n");
+    DEBUG_PUTS("i2c.c: Wait for device to be ready");
     _syncbusy(dev);
 
     /* Set action to ACK. */
     dev->CTRLB.reg &= ~SERCOM_I2CM_CTRLB_ACKACT;
 
     /* Send Start | Address | Write/Read */
-    DEBUG("Generate start condition by sending address\n");
+    DEBUG("i2c.c: Generate start condition by sending address 0x%04"PRIu16"\n",
+            addr);
     dev->ADDR.reg = addr;
 
-    /* Wait for response on bus. */
-    if (addr & I2C_READ) {
-        /* Some devices (e.g. SHT2x) can hold the bus while
-        preparing the reply */
-        if (_wait_for_response(dev, 100 * SAMD21_I2C_TIMEOUT) < 0)
-            return -ETIMEDOUT;
-    }
-    else {
-        if (_wait_for_response(dev, SAMD21_I2C_TIMEOUT) < 0)
-            return -ETIMEDOUT;
-    }
+    /* Wait for response on bus.
+     * Some devices (e.g. SHT2x) can hold the bus while
+     * preparing the reply. */
+    uint32_t timeout = (addr & I2C_READ) ? 100 * SAMD21_I2C_TIMEOUT : SAMD21_I2C_TIMEOUT;
+    int res = _wait_for_response(dev, timeout);
+
+    /* Ignore dev->INTFLAG.bit.ERROR as it does not seem to get set
+     * in the current configuration of the SERCOM module.
+     */
+    bool any_error = dev->STATUS.reg & (
+            SERCOM_I2CM_STATUS_SEXTTOUT /* XXX: do we need this? */ |
+            SERCOM_I2CM_STATUS_ARBLOST | SERCOM_I2CM_STATUS_BUSERR
+            /* not included as always active (hardware bug?) in the
+             * current configuration:
+             * SERCOM_I2CM_STATUS_MEXTTOUT, SERCOM_I2CM_STATUS_LOWTOUT
+             *
+             * not included as only useful with DMA:
+             * SERCOM_I2CM_STATUS_LENERR
+             */
+        ) || (
+            /* bus state should not be unknown at this point */
+            (dev->STATUS.reg & SERCOM_I2CM_STATUS_BUSSTATE_Msk) == BUSSTATE_UNKNOWN
+        ) || (
+            /* we don't handle multi-master layouts and treat this as error */
+            (dev->STATUS.reg & SERCOM_I2CM_STATUS_BUSSTATE_Msk) == BUSSTATE_BUSY
+        ) ||
+        (
+            /* slave busy or address not on bus */
+            dev->STATUS.reg & SERCOM_I2CM_STATUS_RXNACK
+        ) || res;
```

Comment:
Due to `dev->STATUS` being `volatile`, this will force the compiler to re-load that register for every check. Saving that into a temporary variable would allow the compiler to allocate this in a register and keep accessing that register for the bit checks without re-loading the register.

Reply:
In my last commit I copy the register for read access.